### PR TITLE
Minor stylistic changes and refactors

### PR DESCRIPTION
**Pull request by ulugbekna**
- Added a .gitignore file 
- Ran ocamlformat (instead of ocp-indent) to enforce consistent coding style. Hence, the large diff. (Should probably add a git-hook)
- Refactored slightly `run` and `nativeByteEquivalence` functions

### CODE REVIEW

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: While we are in refactoring mood, would you try rewriting those printing functions with `Printf.bprintf`? For example, this case could become

```ocaml
let bound_var sb (x, t) =
  if not typeannot
  then Printf.bprintf sb "%s" x
  else Printf.bprintf sb "(%s : %a)" x typeToOcaml t in
Printf.bprintf sb "fun %a -> %a"
  bound_var (x, t)
  exp m
```

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: This would be clearer rewritten with `match unify_list [ r, t ] with exception No_solution -> ... | ...`.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: This would be clearer with the `None` case first (in matches and if-then-else, I like to always have the shorter case first, instead of being lost after a long blob of code).

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: I would prefer to use the commented code instead of `| _, _`, so that adding a constructor to the datatype results in an exhaustivity warning.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: I would prefer to use an exhaustive list of possible constructors on the first parameter, as suggested for `unify_list` above.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: This would be more readable with `match ... with | s -> ... | exception Not_found -> ...`

File: src/effmain.ml
Comment (gasche) [inline src/effmain.ml]: The placement of `;;` is not so nice. Could you rewrite those expressions into

```ocaml
let () = resetvar ()
let () = resettypevar ()
```

instead? (Consistently using `let () = <expr>` instead of `<expr>` means that double-semicolon are never necessary.)

File: src/effstat.ml
Comment (gasche) [inline src/effstat.ml:10]: I don't understand why `;;` is inserted here, and cannot reproduce with my ocamlformat version (0.9). Which version are you using?

Comment (gasche) [reply src/effstat.ml:10]: It may be that `;;` are including for all items if `;;` were present for some phrases in the module. In that case I would recommend removing all double-semicolon, using the `let () =` trick below.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: These blank lines are an artefact of the handling of documentation-comment (those starting with `(**` instead of `(*` in ocamlformat). Could you change this comment block to use `(* ` instead, so that it gets formatted without extra newlines?

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml:50]: My version of ocamlformat does not break those sequences of commands by default (and I would prefer the not-broken version).

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml:65]: Personally I find the absence of parentheses around the second function unfortunate, but there is no way to fix this with the current ocamlformat. So be it...

**Comment by ulugbekna** [comment]
Having double semicolon after let bindings can come off as a controversial decision, but as it is noted [here](https://discuss.ocaml.org/t/why-does-jane-streets-style-guide-use-double-semicolons/2933), double semicolons are said to help get better errors. 

Also, I used command `ocamlformat -p janestreet -i src/filename.ml` just in case.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml:241]: I do not get this formatting with ocamlformat by default, but only if I use `--break-cases=toplevel`. I think this behavior is a good setting.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: On my ocamlformat version, I only get this behavior if I specify `--single-case=sparse`, and I slightly prefer the default behavior `--single-case=compact`.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: This would be more readable with one item per line, but I haven't found a way to force `ocamlformat` to do it. (The two modes of `break-collection-expressions` are `wrap` or `fit-or-vertical`, and none forces systematic breaking; in particular even using a local `[@@ocamlformat]` attribute does not help). On the other hand, even the non-broken list is sensibly more readable with `--parens-tuple=always`, which I would therefore recommend as the default.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: I would prefer using `--break-separators=after`. My favorite list style is to have either the list on one line, or to have each bracket on its own line, and then one element per line, with a terminating `;` after each item, including the last. I believe this is the best style because it maximizes readability of diffs: adding or removing an item anywhere in the list includes just the new item in the diff. (With the present style, adding a new item at the front also adds a new `;` on the element that was previously in front, so the diff has two line; without the trailing separator, adding an item at the end has the same problem.)

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: I would prefer to use `--no-break-infix-before-func` which was designed specifically to get `varGen >>= fun x ->` here, which is the more readable style once you understand the correspondence with `monadic-let x = varGen in`.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml:699]: The comment on this line is not terribly readable. Maybe you could try to move it before the `in` to force it on the previous line? (I have seen a few similar cases.)

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: Again, this code is sensibly more readable with `--no-break-infix-before-func`.

**Review by gasche — COMMENTED** [review]
I like using automation for style consistency. I noted a few places where the current formatting style is not great, and I marked a few ideas for additional minor refactorings in case you are interested (of course, feel free to also leave the code as it is).

The most important high-level feedback is: there is a clear risk that `ocamlformat` behaves differently on our machines, and that would be bad. I think we have to do two things to minimize the risk:

- We must have a `.ocamlformat` file somewhere (I would suggest the root of the project) that codifies at least all the rules we have explicitly decided, and possibly even all the settings (even the ones which are kept defaults) if that reduces changes between ocamlformat versions. (I would start with just the settings we decided to change explicitly.)

- It would be nice to codify the version dependency on ocamlformat somewhere. The natural place to put this would be an `efftester.opam` file in the toplevel directory, but this file does not exist yet. I am happy to write a version of that file, but maybe you are interested in doing it for learning?

(Eventually we could think of a `dune` port as well; the current build system is minimal and just fine, but `dune` is the fashionable option these days and it does bring a few benefits even for simple projects (`dune utop`, etc.).)

To summarize: could you include an `.ocamlformat` file in your PR with some explicit settings for options that we discussed? It's fine if you disagree with some of the choices I propose, but I would like to be able to check that whatever settings we decide to go with are codified in the file. In particular, once `.ocamlformat` is in, and before merging, I will check that I get the same result with my ocamlformat version.

**Comment by gasche** [comment]
Maybe it could be nice to have an `ocamlformat` or `format` Makefile target that reformats all source file under our agreed ocamlformat style, so that no one (especially occasional contributors) has to wonder how exactly we configured it to reformat after a local change.

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml]: Absolutely! I have been thinking of the same!

**Comment by ulugbekna** [comment]
Thank you so much for a careful review! If that is fine with you, I will proceed with the following to-do list: 

1. implement refactorings that you proposed
2. add `.ocamlformat` to ensure consistency 
3. create efftest.opam to include all dependencies of the project, including the ocamlformat
4. port the project to Dune (it's nice that I have that experience from our work on mlinterp) 
5. move on to comments from pr #2 

Feel free to reorder the list as you find it more appropriate. Also, what do you think about setting up a project board or trello board for this repo?

**Comment by gasche** [comment]
Your proposal is fine with me. My first impulse was to recommend getting the reformatting out of the door first, but in fact given that it is mostly automatic it is completely fine to make other changes first, as your ordering suggests.

File: src/effstat.ml
Comment (ulugbekna) [reply src/effstat.ml:10]: My ocamlformat version is also 0.9. I am using the option `profile=janestreet`, which inserts double semicolons after each phrase.

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml:50]: what sequence of commands do you have in mind? `print_string "."; flush stdout; res`?

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml:65]: I agree. Parentheses would feel more natural there.

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml]: This is how I feel too. Exhaustiveness makes adding types easier.

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml]: I am not sure what you mean here

Comment (gasche) [reply src/efftester.ml]: This is not a formatting comment but more of a refactoring comment: whenever you see a pattern of the form

```ocaml
try
  (* some amount of code *)
with exn -> (* exception handling code *)
```

it is useful to wonder: in the `try .. with` block, what part of the code may actually raise the exception being handled? If there are several different parts, or if it's the whole expression that may raise, then `try .. with` is the right construct to use. If there is only one sub-part of the code that may raise the exception, then it is usually nicer to rewrite the code as

```ocaml
match <part that may raise> with
| exception exn -> (* the exception-handling code *)
| result -> (* the rest of the code *)
```

In the present case, this would suggest the following:

```ocaml
match TypeMap.find key map with
| exception Not_found -> ...
| s -> ...
```

(On the ordering of the two clauses (the value case, the exception case), I usually put whichever is shortest first, with a preference to the exceptional case in case of tie.)

File: src/efftester.ml
Comment (gasche) [reply src/efftester.ml:50]: Yes, but looking back at this I think that this is rather anecdotal. I would be happy with the current formatting.

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml]: Adding this change surely made the code more concise but also more complex (?)

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml]: That indeed makes code more readable. I haven't been familiar with this OCaml feature before. Thanks! :-)

**Comment by ulugbekna** [comment]
I think I am done with refactoring review comments. I think we can keep an open eye for additional changes along the way.

I will move on to adding an `ocamlformat` file to the root of the repo now. If there are any more refactoring comments, I am glad to discuss and implement them.

File: src/effstat.ml
Comment (ulugbekna) [reply src/effstat.ml:10]: I think `let-binding-spacing = double-semicolon` is set in the janestreet profile. I am not opinionated on this and can change to something else if you want. It was noted [here](https://discuss.ocaml.org/t/why-does-jane-streets-style-guide-use-double-semicolons/2933), however, that double semicolons can improve the error messages

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: You are right that this code is more complex than what was there before, but this can be improved. The key is to move the conditional behavior to be as local as possible. Currently you are testing (1) whether the arrow should be effect-annotated or not and (2) whether the left-hand-type should be parenthesized or not, bot at the toplevel, which makes 4 cases to handle (corresponding to the four format strings in your code). With the right auxiliary definitions, you could write the following instead:

```ocaml
Printf.bprintf "%a -%a> %a"
  print_simple_type s
  print_effannot (if effanot then Some e else None)
  print_type t
```

A type expression is "simple" if it is clearly delimited, either because it is a single word, or because it starts and ends with matching delimiters. `print_simple_type` prints any type expression as a simple type: in the `Fun _` case, it adds parentheses around the type.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: Nitpick: I would recommend to not mix `Printf` and `Buffer` functions for consistency, using `Printf.bprintf sb "string"` in that case.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: This function should go as an auxiliary definition above.

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]: With the right auxiliary definitions:

```ocaml
Printf.bprintf sb "fun %a -> %a"
  print_binder (x, t)
  exp m
```

File: src/efftester.ml
Comment (gasche) [inline src/efftester.ml]:
```ocaml
Printf.bprintf sb "let %a = %a in %a"
  print_binder (x, t)
  exp m
  exp n
```

**Review by gasche — APPROVED** [review]
Thanks, this is looking very nice right now.

Since you asked, I made a last round of nitpicky suggestions. But essentially I think that the PR is good as it is now; feel free to make some further refactoring changes if you wish, but generally I think this is good to go (let me know if/when you agree, and I'll merge).

I think that the move to Dune can go in a separate PR, to limit the scope of this one to reformatting + refactorings. (Maybe we should have not mixed the two in the first place.)

File: src/efftester.ml
Comment (ulugbekna) [reply src/efftester.ml]: That's a nice solution. The function application, however, looks like this 
```ocaml
Printf.bprintf
      sb
      "%a -%a> %a"
      print_simple_type
      s
      print_effannot
      (if effannot then Some e else None)
      print_type
      t
``` 
When the new version of ocamlformat is published, it will have an option `break-fun-sig`, which might solve this problem.

**Comment by ulugbekna** [comment]
I think I resolved problems pointed in the latest comments. 

Regarding the `efftester.opam`, shall we include it to this pr?

**Comment by gasche** [comment]
I would be in favor of pushing as-is, because I think that both the reformatting and the refactoring are satisfying. Merging things is making progress! You can send another PR for the opam file.

On the other hand, from this branch `./effmain.native` does not work anymore, because of an issue somewhere in the commit https://github.com/ulugbekna/efftester/commit/9372196a5a8f3f73ed9440b9694053cddf1cc875. We need to fix this bug before merging the PR.

(If you know how to use interactive-rebase to amend commits in place, that would be nice for the bugfix to this commit, so that the intermediate states within the PR are all correct.)

**Comment by ulugbekna** [comment]
Working on fixing the bug.

**Comment by jmid** [comment]
> Working on fixing the bug.

A suggestion: comment in `ocaml_test` in `effmain.ml`. It's purpose is precisely to test whether the generated program passes the OCaml compiler's type check.

File: src/efftester.ml
Comment (jmid) [inline src/efftester.ml:1373]: I suggest putting the comment within the last parenthesis.
The point is to highlight where the goal is: "generate an int program which can have an effect".
If the comment follows two lines after, that meaning is lost.

**Comment by gasche** [comment]
> A suggestion: comment in ocaml_test in effmain.ml. It's purpose is precisely to test whether the generated program passes the OCaml compiler's type check.

Indeed. One thing that would be nice, in the medium term (not terribly urgent but nice-to-have), would be to have a test target that runs the sanity-check tests (which are currently commented out). As @ulugbekna is planning to use Dune, it's better to wait after dune to not have to do the build-system part twice.

**Comment by ulugbekna** [comment]
Wrongly typed programs have more arguments than necessary, e.g. `int_of_string exit 0`. We can see that the problem is in parentheses and if we put them appropriately we get `int_of_string (exit 0)`. Hence, if we put parentheses in `app` function's format string as follows, I cannot reproduce the bug anymore. 
```ocaml
  and app sb t =
    match t with
    | App (_, m, _, n, e) -> Printf.bprintf sb "(%a %a)" app m app n
    | _ -> arg sb t
```
However, this seems to put unnecessary parentheses in the outer-most function application. Let me see if there is something else I am missing.

**Comment by gasche** [comment]
One small thing missing in your analysis is the place where the bug was introduced. I bisected the change and I know it comes from your commit https://github.com/ulugbekna/efftester/commit/9372196a5a8f3f73ed9440b9694053cddf1cc875. Before that commit, the code was working, and in it you changed something that introduced this bug. One way to fix the bug is to devise a new fix as you are proposing, but another is to locate the (faulty) change in behavior and revert it. I believe that the second approach is not too hard now that you did the analysis work, and probably doesn't suffer from the same disadvantage.

**Comment by ulugbekna** [comment]
AFAIU, the bug is fixed. The problem was as expected in `app` function where I did a minor mistake in refactoring: replaced function `arg` with `app`. Moving to Dune and introducing some checks/tests would indeed prevent such mistakes in future. 

If there are no comments on this, I will switch back to working on introducing opam and dune.

**Comment by gasche** [comment]
Merged, thanks!

**Pull request merged**
